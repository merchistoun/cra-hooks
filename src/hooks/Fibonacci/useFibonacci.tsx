import { useSuffix } from "../";

export interface UseFibonacci {
  fibonacci: (n: number) => string[];
}

export const useFibonacci = (): UseFibonacci => {
  const { addSuffix } = useSuffix();

  const fibonacci = (n: number): string[] => {
    const arr = new Array(n);

    for (let i = 0; i < n; i++) {
      arr[i] = i > 1 ? arr[i - 1] + arr[i - 2] : 1;
    }

    return arr.map(addSuffix);
  };
  return { fibonacci };
};
