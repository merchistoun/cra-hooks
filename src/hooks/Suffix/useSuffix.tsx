export interface UseSuffix {
  addSuffix: (value: number) => string;
}

export const useSuffix = (): UseSuffix => {
  const addSuffix = (value: number): string => {
    switch (value % 10) {
      case 1:
        return `${value}st`;
      case 2:
        return `${value}nd`;
      case 3:
        return `${value}rd`;
      default:
        return `${value}th`;
    }
  };

  return { addSuffix };
};
