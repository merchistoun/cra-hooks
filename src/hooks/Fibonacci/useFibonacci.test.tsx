import { UseFibonacci, useFibonacci } from "./useFibonacci";
import { useSuffixMock as mockUseSuffix } from "../Suffix/useSuffix.mock";
import { renderHook } from "@testing-library/react";
import { it, describe, expect } from "@jest/globals";

afterEach(() => {
  jest.clearAllMocks();
});

beforeEach(() => {
  jest.clearAllMocks();
});

const mockDependencies = (overrides = {}) => {
  return {
    useSuffix: mockUseSuffix(overrides),
    ...overrides,
  };
};

describe("useFibonacci hook", () => {
  it("should mount without error", async () => {
    const { useSuffix } = mockDependencies({
      addSuffix: jest.fn().mockImplementation((i) => `${i}`),
    });

    let hook: UseFibonacci;
    const { result } = renderHook(() => useFibonacci());
    hook = result.current;
    expect(hook).toMatchSnapshot();

    const { fibonacci } = hook;

    expect(fibonacci(5)).toEqual(["1", "1", "2", "3", "5"]);
    expect(useSuffix.addSuffix).toBeCalledTimes(5);
  });
});
