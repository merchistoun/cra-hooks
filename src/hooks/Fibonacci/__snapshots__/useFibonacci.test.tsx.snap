// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFibonacci hook should mount without error 1`] = `
Object {
  "fibonacci": [Function],
}
`;
